### PR TITLE
add functions to change timeout

### LINKED ISSUE #26: Extending the timeout
So far as I understand it, the timeout value is set when you add the `connect-timeout` middleware when initializing the express server, so the same timeout applies to every request the comes in thereafter (that matches the route concerned).

What I'm finding is that for some incoming GET requests, there is a particular combination of options in the request that causes a network fetch to a remote resource than is not very reliable.  For those requests, I'd like to relax the timeout value for `connect-timeout` somewhat, to allow for a few fetch retries.

But I can't see any way to change the timeout value 'on the fly', for a particular request.  The only control I can see over the timeout is to use `req.clearTimeout()` whilst processing a request.  This is helpful, but `timeout is completely removed and will not fire for this request in the future`.  This is not entirely what I'm looking to achieve... I would like to retain a timeout, but just a slightly longer one.

Is what I'm asking for really naughty, breaking established rules of good practice of which I'm not aware?  Or is it a reasonable thing to want to do?

### PR DESCRIPTION
**Pull request by UziTech**
add functions to `req` to change timeout

`req.resetTimeout(delay)`: reset the timeout and start from now
`req.addTimeout(delay)`: add to the current timeout
`req.getTimeout()`: get the amount of milliseconds left (0 after timedout)

fixes #26

### CODE REVIEW

**Comment by dougwilson** [comment]
Hi @UziTech thanks for the pull request! I'll start reviewing over the coming days. The following jumped out at me just with an initial glance at the pull request, though:

1. There is no documentation. Some folks wait to write documentation after the review, but you didn't mention that in your description, so wasn't sure if you were waiting or accidentally forgot to write it.
2. `req.setTimeout` already exists -- https://nodejs.org/dist/latest-v10.x/docs/api/http.html#http_message_settimeout_msecs_callback . You'll need to use a different method name as not to overwrite core Node.js method and break existing apps (and just otherwise prevent apps from being able to access this functionality).
3. It doesn't seem like `req.clearTimeout()` will stop `req.timeoutLeft()` from continuing to count down as if the timeout was still there.

**Comment by UziTech** [comment]
Good points.

  1. I figured a few things would change so I was waiting on the docs.
  2. What would you like it to be called? A couple suggestions:
    a. `.startTimeout()`
    b. `.restartTimeout()`
    c. `.timeout()`
  3. Nice catch, I will fix that.

**Comment by UziTech** [comment]
also a thought: should `.timeoutLeft()` be changed to `.timeoutRemaining()`

P.S. It's a good thing this problem doesn't involve cache invalidation.

**Comment by UziTech** [comment]
I renamed the functions:

```
req.setTimeout(...) -> req.resetTimeout(...)
req.timeoutLeft(...) -> req.getTimeout(...)
```

and added docs to the README

**Comment by UziTech** [comment]
@dougwilson any chance you got to review this?

**Comment by UziTech** [comment]
🙃

**Comment by UziTech** [comment]
Is anyone looking at this?

/cc @dougwilson @juliangruber @jonathanong

**Comment by juliangruber** [comment]
not at this moment, sorry

**Comment by UziTech** [comment]
You can remove the `need docs` label. I added docs to explain the new methods.

**Comment by UziTech** [comment]
It would be great to get this merged.

/cc @dougwilson @juliangruber @jonathanong

File: index.js [hunk 1]
```diff
@@ -44,19 +44,52 @@ function timeout (time, options) {
...
+        }))
+      })
     }
 
     req.clearTimeout = function () {
+      delay = 0
       clearTimeout(id)
     }
 
+    req.resetTimeout = function (newDelay) {
```
Comment (dougwilson) [inline index.js]: If calling reset after the timeout already occurred, should req.timedout be reset back to false? Or should is stay true and update docs to describe it just represents the first trigger? Or not allow a reset after timeout happened?

**Comment by dougwilson** [comment]
Sorry. I removed the label and left one comment about the reset behavior.

File: test/test.js [hunk 2]
```diff
@@ -106,6 +106,162 @@ describe('timeout()', function () {
...
+          res.end('Hello')
+        })
+
+      request(server)
+      .get('/')
+      .expect(200, 'Hello', done)
+    })
+    it('should reset the timeout with 5000 default', function (done) {
+      var server = createServer(null,
+        function (req, res) { req.resetTimeout() },
```
Comment (dougwilson) [inline test/test.js]: How does this test validate that 5000 is what the reset used? It is identical to the first test just missing the argument.

It seems like if the impl was change to not accept an argument, all three of these tests still pass?

File: index.js (hunk 1 above)
Comment (UziTech) [reply index.js]: it should probably remain false to let the user know that it was triggered.

File: test/test.js (hunk 2 above)
Comment (UziTech) [reply test/test.js]: Good call. I updated the tests to test the actual timeout.

File: index.js (hunk 1 above)
Comment (UziTech) [reply index.js]: It remains `true` if `req.clearTimeout()` is called after timeout is fired.

I updated the docs to reflect this.

**Comment by UziTech** [comment]
@dougwilson this should be good to go.

**Comment by UziTech** [comment]
@dougwilson can this get reviewed?

**Comment by UziTech** [comment]
It would be great to get this merged.

/cc @dougwilson @juliangruber @jonathanong

**Comment by dougwilson** [comment]
Hi @UziTech sorry. There are still a few issues with the PR and I was just going to resolve them for you instead of doing more back and forth, I just haven't gotten to it yet.

**Comment by UziTech** [comment]
@dougwilson can we get this merged?

**Comment by UziTech** [comment]
@dougwilson are you still able to merge this?

**Comment by UziTech** [comment]
Over a year and a half waiting for this PR to be merged!

@expressjs If you need help maintaining some of the expressjs packages I would be willing to help.

**Comment by UziTech** [comment]
We are almost at the 2 year mark. Is there anything I can do to move this along?

**Comment by dougwilson** [comment]
Hi @UziTech , I think that with a new sweeping overview we are doing, this module is likely to just be discontinued and deprecated, as it will not work at all with the upcoming Express, as it calls `next()` more than once and other design issues. I think we should just be real at this point: if this is something you need / want, just publish the module as a fork and go from there; even the name of this module is problematic for our organization, having the `connect-` prefix.

**Comment by UziTech** [comment]
Thanks for the response.

**Comment by dougwilson** [comment]
And @UziTech I hope you can accept my apology for letting this sit here -- I really didn't want to continue this module in many ways and the original authors didn't seem to be around responsive, either. If it does get a fresh breath of air, having the module not start with `connect-` would be awesome anyhow, as it makes little sense with Express since it hasn't been built on connect in years... I do have a couple legacy apps that use this module still, but there are design issues that makes it hard to use; the readme shows putting a "stopper" middleware between everything to at least make it _kind of_ work, but it... sucks :(

It may make sense to rebirth as like a middleware _wrapper_ instead. To do something like `app.use(timeout(200, myMiddleware))` that would time out that middleware at 200ms, IDK. I haven't put too much thought into it, and I think that is part of the problem--it deserves to be in the hands of someone who can really put thought into it...

**Comment by UziTech** [comment]
No problem. I just want to make clear that, whatever timeout method is used, it is important to have the ability to cancel/extend the timeout in certain situations.

In my project a middleware wrapper wouldn't work because there are multiple middleware that could cause a timeout, so it works better to set a timeout over the entire request.

As I said above, If you need any help maintaining small projects I would be happy to help 😁 :+1:

**Pull request closed without merging**